packages/server/lib/files.js

```diff
@@ -2,45 +2,93 @@ const path = require('path')
 const { fs } = require('./util/fs')
 
 module.exports = {
-  readFile (projectRoot, file, options = {}) {
+  async readFile (projectRoot, file, options = {}) {
     const filePath = path.resolve(projectRoot, file)
     const readFn = (path.extname(filePath) === '.json' && options.encoding !== null) ? fs.readJsonAsync : fs.readFileAsync
+    const readFileAbortController = new AbortController()
 
     // https://github.com/cypress-io/cypress/issues/1558
     // If no encoding is specified, then Cypress has historically defaulted
     // to `utf8`, because of it's focus on text files. This is in contrast to
     // NodeJs, which defaults to binary. We allow users to pass in `null`
     // to restore the default node behavior.
-    return readFn(filePath, options.encoding === undefined ? 'utf8' : options.encoding)
-    .then((contents) => {
+
+    const readOptions = {
+      encoding: options.encoding === undefined ? 'utf8' : options.encoding,
+      signal: readFileAbortController.signal,
+    }
+
+    let readFileTimeout
+
+    if (options.timeout !== undefined) {
+      readFileTimeout = setTimeout(() => {
+        readFileAbortController.abort()
+      }, options.timeout)
+    }
+
+    try {
+      // graceful-fs does not support the standard fs Promise API.
```

Comment:
What benefit is graceful-fs giving us?

Reply:
mainly, `graceful-fs` automatically queues/retries on `EMFILE` errors, which is the error thrown when there are too many open file handles. occasionally Cypress hits this (so I've been told, I've never seen it), so that's why we use graceful-fs.

That being said... graceful-fs seems to be no longer maintained, so we should probably move away from it at the first excuse we get.